docs/proposals/templates.md

```diff
@@ -0,0 +1,349 @@
+          name: "$(DATABASE_SERVICE_NAME)"
+    - 
+      kind: "ReplicationController"
+      apiVersion: "v1"
+      metadata: 
+        name: "$(DATABASE_SERVICE_NAME)"
+      spec: 
+        replicas: "$(REPLICA_COUNT)"
+        selector: 
+          name: "$(DATABASE_SERVICE_NAME)"
+        template: 
+          metadata: 
+            creationTimestamp: null
+            labels: 
+              name: "$(DATABASE_SERVICE_NAME)"
+          spec: 
+            containers: 
+              - 
+                name: "mongodb"
+                image: "docker.io/centos/mongodb-26-centos7"
+                ports: 
+                  - 
+                    containerPort: 27017
+                    protocol: "TCP"
+                env: 
+                  - 
+                    name: "MONGODB_USER"
+                    value: "${MONGODB_USER}"
+                  - 
+                    name: "MONGODB_PASSWORD"
+                    value: "${MONGODB_PASSWORD}"
+                  - 
+                    name: "MONGODB_DATABASE"
+                    value: "${MONGODB_DATABASE}"
+```
+
+## Template Expansion API
+
+A server side endpoint for expanding templates is proposed.  The endpoint will inspect the `templateKind` for the template object
+being posted to the endpoint, and then use that information to determine the appropriate template expansion plugin to invoke.
+
+A template expansion plugin is expected to return a `List` of valid k8s objects, which the template processing endpoint
+will then return to the original caller.  This represents the primary contract for templates:  A template api object 
+envelope containing arbitrary content is converted into a list of k8s objects.
+
+The actual mechanism for looking up and invoking a template expansion plugin (as well as for how one is provided to k8s) is
+still being defined.  One suggestion was to essentially run a template expansion plugin as a microserver on the k8s cluster
+itself, possibly within a namespace with a name that matched the `templateKind`.  However dictating specific namespaces
+be available within the cluster for a plugin feels overly burdensome.  A map between the `templateKind` and
+the api endpoint capable of expanding it is necessary, however.
+
+Template processors could also be provided in client form to help template authors perform local testing+debugging.
+
+Nested templates should be handled by the primary endpoint, retrieving and handing each nested template off to the appropriate
+expansion plugin, and then assembling the aggregate list of objects from the top level template and any nested templates.
+
+## Open discussion points
+### Where to define parameters (closed?)
+
+There has been some discussion around where to define parameters that are being injected into a Template
+
+1. In a separate standalone file
+2. Within the Template itself
+
+This proposal suggests including the parameter definitions within the Template, which provides a self-contained structure that
```

Comment:
I was originally thinking we would also need to ability to call out to a file on the side.  But having read this, I now think there may be no need to allow the "envelope" to refer to a secondary file, because for all languages where this makes sense, the language has a way to import a second file itself.  So for the Python form:

body: |
  import foo

would be equivalent to

bodyFile: "foo.py"

Similarly Jinja has {% include %}, Jsonnet has import "foo.jsonnet", etc.  In all cases, one can just use a small snippet of the template language's own code to invoke the file.